docs/backend/deprecation.md

```diff
@@ -0,0 +1,314 @@
+
+While refactoring code, moving modules, functions, classes and methods is often needed.
+To not break third party code imports from the old place or usage of old functions/ methods must work for while.
+Deprecated methods are usually removed with the next major release of Plone.
+
+Following the [semantic versioning guideline](http://semver.org) is recommended.
+
+### Help Programmers, No annoyance
+
+Deprecation has to support the consumers of the code - the programmers using it.
+From their point of view, Plone core code is an API to them.
+Any change is annoying to them anyway, but they feel better if deprecation warnings are telling them what to do.
+
+Deprecations must always log at level *warning* and have to answers the question:
+
+**"Why is the code gone from the old place? What to do instead?"**
+
+A short message is enough., i.e.:
+
+- "Replaced by new API xyz, found at abc.cde".,
+- "Moved to xyz, because of abc.",
+- "Name had a typo, new name is "xyz".
+
+All logging has to be done once, i.e. on first usage or first import.
+It must not flood the logs.
+
+### Use Cases
+
+Renaming
+
+: We may want to rename classes, methods, functions or global or class variables in order to get a more consistent api or because of a typo, etc.
+  We never just rename, we always provide a deprecated version logging a verbose deprecation warning with information where to
+  import from in future.
+
+Moving a module, class, function, etc to another place
+
+: For some reason, i.e. merging packages, consistent api or resolving cirular import problems, we need to move code around.
+  When imported from the old place it logs a verbose deprecation warning with information where to import from in future.
+
+Deprecation of a whole package
+
+: A whole package (folder with `__init__.py`)
+
+  - all imports still working, logging deprecation warnings on first import
+  - ZCML still exists, but is empty (or includes the zcml from the new place if theres no auto import (i.e. for meta.zcml).
+
+Deprecation of a whole python egg
+
+: We will provide a last major release with no 'real' code, only backward compatible (bbb) imports of public API are provided.
+  This will be done the way described above for a whole package.
+  The README clearly states why it was moved and where to find the code now.
+
+Deprecation of a GenericSetup profile
+
+: They may got renamed for consistency or are superfluos after an update.
+  Code does not need to break to support this.
+
+## Enable Deprecation Warnings
+
+### Zope
+
+Zope does configure logging and warnings, so the steps below (under section Python) are not needed.
+
+Using `plone.recipe.zope2instance` add the option `deprecation-warnings = on` to the buildouts `[instance]` section.
+
+```ini
+[buildout]
+parts = instance
+
+[instance]
+recipe = plone.recipe.zope2instance
+...
+deprecation-warnings = on
+...
+```
+
+This just sets a configuration option in `zope.conf`.
```

Comment:
~~~suggestion
This adds this line to the `zope.conf` file:

```
debug-mode on
```
~~~

Or maybe add these lines above and then say that buildout user can have this with the proper recipe settings.